### ISSUE TITLE
Store server's version number, include in Sentry events

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#3745]
When an error happens only on certain versions of the Zulip server, that can be a powerful clue for diagnosing it. I've often noticed when looking at Sentry events that I'd like to have that information. #3732 is a recent example that brought this to mind.

In implementing this, one layer is to use the Sentry API to get that information attached. I'm not sure how best to do that; @ray-kraesig has studied those docs recently and may have ideas.

The other layer is to get and keep the server-version information around in the first place. The server reports its version number in the `server_settings` response, but we don't currently store it anywhere; the only things we use the `server_settings` response for are at initial login, so we fetch it once then. It wouldn't do to just store it then, either, because of course the version (and other parts of that response) can change over the months and years the user might stay logged in.

Here's a plausible approach:
* Add a property like `serverVersion` to the `Account` type, found in the `accounts` state-subtree. Can be `null | string`.
* In the login flow, save the version we get.
* At the same time we call `/register` aka `registerForEvents`, in `doInitialFetch` in `fetchActions.js`, make a `/server_settings` request again and save the new version value.

The server version number could also be useful if we decide to condition on it for certain API changes, as suggested at https://github.com/zulip/zulip-mobile/issues/3732#issuecomment-569468538 .

**Comment by ray-kraesig** [comment]
> In implementing this, one layer is to use the Sentry API to get that information attached. I'm not sure how best to do that; @ray-kraesig has studied those docs recently and may have ideas.

Sounds like a job for either [tags](https://docs.sentry.io/enriching-error-data/context/?platform=javascript#tagging-events) or [extras](https://docs.sentry.io/enriching-error-data/context/?platform=javascript#extra-context). We probably want something like this when handling the `server_settings` response:
```js
Sentry.configureScope(scope => { scope.setTag("server-version", "2.2.7723"); });
```

Note that I haven't tried tags out yet, so I'm not sure what they look like in Sentry – but extras support is used in #3733.

**Comment by gnprice** [comment]
> Sounds like a job for either [tags](https://docs.sentry.io/enriching-error-data/context/?platform=javascript#tagging-events) or [extras](https://docs.sentry.io/enriching-error-data/context/?platform=javascript#extra-context).

Thanks!

Between those two, I agree with your example that we probably want tags. Seems like the main difference is that tags get indexed, searched by, and aggregated on -- all of which sound like useful things for the server version.

Thinking about the search and aggregation, one thing that occurs to me is we likely want several tags here:
* the full version string, for complete information (e.g. `2.1.0-66-gd8a2ec0`, the current value of `curl -s https://python.zulipchat.com/api/v1/server_settings | jq .zulip_version -r`)
* one or two simplified version strings, making it more coarse-grained for aggregation. (e.g. `2.1`, or for explicitness perhaps `2.1.x`)

**Comment by ray-kraesig** [comment]
One reason I'm not sure we want tags is because they may also get *dis*aggregated on – that is, a truly client-side-only issue might end up separated into fifty different issues just because the clients exhibiting it are connected to fifty different server versions.

(The links at the top of [this documentation](https://docs.sentry.io/data-management/event-grouping/) describe ways to customize the grouping, though, so even if that does happen we can probably override it.)

**Comment by gnprice** [comment]
Mmm, I see. I *think* that's not what will happen because the examples include:
> The version of your platform (e.g. iOS 5.0)

and so I expect it to behave like the existing fields of that flavor do: they don't affect how the events get grouped into issues, and instead they get counted up across all the events of an issue to produce the graphs seen at the RHS of an event page.

But yeah, if we get surprised in that respect then we'll figure out a way to deal. :slightly_smiling_face:

**Comment by chrisbobbe** [comment]
I can make a PR with the following —

```
* Add a property like `serverVersion` to the `Account` type, found in the `accounts` state-subtree. Can be `null | string`.
* In the login flow, save the version we get.
* At the same time we call `/register` aka `registerForEvents`, in `doInitialFetch` in `fetchActions.js`, make a `/server_settings` request again and save the new version value.
```

— and then the other things that depend on the server version (Sentry logging, etc.) can follow, right?

How's this for a workflow: I'll assign this issue to me, submit the PR without a `Fixes:` mark, and un-assign this issue so the rest of the stuff can be done. (I'm happy to do it but it sounds like we have more discussion to do about Sentry, etc.)

First, is there a reason not to just store the entire object received from server_settings?

**Comment by chrisbobbe** [comment]
It seems like the server_settings data (whether just the server version or the whole object) would make sense if made part of both the REALM_ADD and REALM_INIT actions; does this seem right conceptually? The alternative is most likely to create a new action.

**Comment by chrisbobbe** [comment]
OK, so from our chat earlier:

> First, is there a reason not to just store the entire object received from server_settings?

Yes, we don't need all that stuff. Just the version number is fine.

> It seems like the server_settings data (whether just the server version or the whole object) would make sense if made part of both the REALM_ADD and REALM_INIT actions; does this seem right conceptually? The alternative is most likely to create a new action.

Right, those seem to make sense.

**Comment by ray-kraesig** [comment]
> OK, so from our chat earlier:
> 
> > First, is there a reason not to just store the entire object received from server_settings?
> 
> Yes, we don't need all that stuff. Just the version number is fine.
> 

Maybe not all of it, but we do want some of them in support of other issues. `realm_icon` is needed for #392, and `realm_uri` is needed to properly fix... er, the issue that #3671 is intended for.

**Comment by gnprice** [comment]
> Maybe not all of it, but we do want some of them in support of other issues. `realm_icon` is needed for #392, and `realm_uri` is needed to properly fix... er, the issue that #3671 is intended for.

True! I think we don't want to keep the bulk of it, though (which stops being relevant after we log in.) So I think it'll make most sense to pick out the bits we want selectively, as we add code that wants to use them.

**Comment by chrisbobbe** [comment]
> I think we don't want to keep the bulk of it, though

This risks becoming a tangent, but we do at least pass the entire object in `dispatch(navigateToAuth(serverSettings));` (RealmScreen.js)

**Comment by chrisbobbe** [comment]
> This risks becoming a tangent, but we do at least pass the entire object in dispatch(navigateToAuth(serverSettings)); (RealmScreen.js)

(As noted in-person, this is a one-off use of it and not really a concern.)

**Comment by gnprice** [comment]
Bumping priority because this is on the path to resolving #3732.

**Comment by gnprice** [comment]
#3839 is merged, so we are now storing -- and parsing -- the server version!

**Comment by gnprice** [comment]
We've resolved #3732 thanks to using this information. But I think the remaining part is also a priority: when looking at Sentry issues, I'd really like to be able to see what Zulip server versions were involved.

(Probably I should have filed this as two separate issues in the first place.)

I think the next steps are to add some Sentry tags, in the way Ray describes in [the first comment](https://github.com/zulip/zulip-mobile/issues/3745#issuecomment-569989760). Specifically, as I added after that:

> Thinking about the search and aggregation, one thing that occurs to me is we likely want several tags here:
> 
> * the full version string, for complete information (e.g. `2.1.0-66-gd8a2ec0`, the current value of `curl -s https://python.zulipchat.com/api/v1/server_settings | jq .zulip_version -r`)
> * one or two simplified version strings, making it more coarse-grained for aggregation. (e.g. `2.1`, or for explicitness perhaps `2.1.x`)

**Comment by chrisbobbe** [comment]
https://github.com/zulip/zulip-mobile/pull/3952 is on the path to this, I think, right? I believe I've addressed the most recent review, but I think there was another strategy we were considering as an alternative to the strategy used in that PR.

**Comment by ray-kraesig** [comment]
A new selector `getServerVersion` was added in #4022 (commit 11cca9b3b8d132f8ebd32c82c725a2832314b59e). I think that's enough to keep #3952 from blocking this.

**Comment by gnprice** [comment]
Ah, I'd forgotten about that PR, sorry. Yeah, I think it's orthogonal to starting to send the versions in Sentry events.

**Comment by gnprice** [comment]
Bump -- [when discussing](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.23M4156.20Message.20List.20placeholders/near/928737) #4156 we were reminded that this would be a good thing to finish.

**Comment by gnprice** [comment]
We [discussed this in chat](https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/.23M3745.20Server.20version.20to.20Sentry/near/930141) back in July, and found some obstacles from the fact that trying to get at our Redux store from the logging module introduces a large tangle of import cycles.

That tangle may get resolved when we complete #3804, removing react-navigation from our Redux state, which is looking soon.

If not, a direct workaround would be to break some of the abstraction layers involved: have a tiny module that imports nothing and just exports a global, and have our reducer update that global any time it's updating the server version in our Redux state.

**Comment by chrisbobbe** [comment]
Now that #3804 is resolved, https://github.com/zulip/zulip-mobile/pull/4191 should bring us within reach of being able to send the server version to Sentry—I just attempted it, and I didn't get any complaints from Flow or Jest! 🙂
